### PR TITLE
Keep ::helper objects alive while in use by helper_servers

### PR DESCRIPTION
**Pull request by eduard-bagdasaryan**
Squid creates ::helper objects to manage a given configured helper. For
each ::helper object, Squid may create one or more helper_server objects
to manage communication with individual helper processes. A
helper_server object uses its so called "parent" ::helper object to
access configuration and for helper process death notification purposes.
There are no checks that the "parent" object is still alive when used.

The same problem applies to statefulhelper and helper_stateful_server.

helper_server code evidently attempted to extend their "parent" lifetime
using cbdata, but that does not work, especially in C++ world where the
object is destructed (and, hence, enters an undefined state) even though
its top-level memory is preserved by cbdata. Non-trivial members like
helper::queue (std::queue) and statefulhelper::reservations
(std::unordered_map) release their internal memory when destructed.

We now refcount ::helper objects to keep them alive until the last
object using them is gone. This does not result in reference loops
because the ::helper object uses raw dlink_node pointers to store its
helper_servers.

The following helpers (listed here by their container names) were
destructed while possibly still in use by their helper_server objects:
external_acl::theHelper, Ssl::CertValidationHelper::ssl_crt_validator,
Ssl::Helper::ssl_crtd. The following helpers are not destructed during
reconfiguration: redirectors, storeIds, basicauthenticators,
ntlmauthenticators, negotiateauthenticators, and digestauthenticators
(even though casual reading of relevant code may suggest otherwise).

This bug fix does not address or mark many remaining helper bugs.

### CODE REVIEW

File: src/auth/basic/Config.cc
Comment (yadij) [inline src/auth/basic/Config.cc:39]: Pointer types should never need explicit ```nullptr``` instantiation.
```suggestion
HelperPointer basicauthenticators;
```

File: src/auth/negotiate/Config.cc
Comment (yadij) [inline src/auth/negotiate/Config.cc:35]:
```suggestion
StatefulHelperPointer negotiateauthenticators;
```

File: src/auth/digest/Config.cc
Comment (yadij) [inline src/auth/digest/Config.cc:49]:
```suggestion
HelperPointer digestauthenticators;
```

File: src/auth/ntlm/Config.cc
Comment (yadij) [inline src/auth/ntlm/Config.cc:36]:
```suggestion
StatefulHelperPointer ntlmauthenticators;
```

File: src/redirect.cc
Comment (yadij) [inline src/redirect.cc:58]:
```suggestion
static helper::Pointer redirectors;
static helper::Pointer storeIds;
```

File: src/helper/forward.h
Comment (yadij) [inline src/helper/forward.h:30]: Please avoid defining multiple different types referring to the same thing. Specifically the ```helper::Pointer``` and ```statefulhelper::Pointer``` types duplicate these in a confuse-able way.

FWIW, my rule of thumb is to prefer the class-scoped definitions whenever possible. However if the code needs global-scope types (as this helper code might), move everything to use those instead. So each class with a ```Pointer``` in public ```.h``` files only has one of the two styles. Type symbols defined locally in a ```.cc``` function do not count for this rule.

Comment (yadij) [inline src/helper/forward.h:30]: If you keep these new types, please move into the ```namespace Helper``` scope.

File: src/helper.h
Comment (yadij) [inline src/helper.h:72]: Please revert the code style to using C++ construction instead of a maker-factory style.

```suggestion
explicit helper(const char *name) : id_name(name) {}
```

File: src/helper.h
Comment (yadij) [inline src/helper.h:145]: Please revert the code style to using C++ construction instead of a maker-factory style.

```suggestion
explicit statefulhelper(const char *name) : helper(name) {}
```

**Review by yadij — CHANGES_REQUESTED** [review]

**Review by eduard-bagdasaryan — COMMENTED** [review]
Partially done at 14593dc.

File: src/helper/forward.h
Comment (rousskov) [reply src/helper/forward.h:30]: IMO, the primary reason we should not move these aliases is because such a move would create a "Foo inside Foo" problem: Helper::HelperPointer does not look good and Helper::Pointer is just wrong. The problem here is in the name of the `Helper` namespace or the name of the `::helper` class (that this PR does not and should not change).

I hope we will rename the `::helper` class eventually. If we can agree on that future class name _now,_ then we can move these new pointers inside the Helper namespace using that new name in _this_ PR (and then rename the classes in another dedicated PR).

I can suggest Helper::Client (for the current `::helper` class) and Helper::Session (for the current `helper_server`). That naming scheme matches what the corresponding classes represent well. Squid already uses similar Client terminology for HTTP and ICAP clients.

That (future) naming scheme would result in _this_ PR using Helper::ClientPointer and Helper::StatefulClientPointer. Another dedicated PR will then rename the other/existing classes and move them into the Helper namespace (and into the src/helper/ directory?). We just need to merge a few existing bug fixes (like this PR) so it should not take long.

@yadij, would you support this new Client/Session naming scheme for helper-related classes?

**Comment by rousskov** [comment]
@yadij, please note that this PR has been waiting for your review for more than 20 days. Under the "Timeouts for stalled PRs" experiment, it may be force-merged in 10 days.

**Comment by rousskov** [comment]
rebuild

**Comment by rousskov** [comment]
> @yadij, please note that this PR has been waiting for your review for more than 20 days. Under the "Timeouts for stalled PRs" experiment, it may be force-merged in 10 days.

Those 10 days also passed. Under the "Timeouts for stalled PRs" [experiment](http://lists.squid-cache.org/private/squid-board/2023-June/001806.html), I am manually dismissing the blocking review and will clear this PR for merging. If there are any objections, please voice them _now_. CC: @yadij, @kinkie.

**Comment by yadij** [comment]
Hmm, I did not get any notifications about this.

File: src/helper.h
Comment (yadij) [reply src/helper.h:145]: > > Amos: Please revert the code style to using C++ construction instead of a maker-factory style.
> 
> Why?! PR code (returning a smart pointer) is safer than raw `new` (returning a pointer that can easily leak).

All callers are immediately and unconditionally de-referencing the pointer - regardless of whether it is raw or smart. That invalidates the claim of their being any difference in safety.

> The new API also directs new helper users to proper pointer storage. As an added bonus, this `X::Make()` upgrade significantly reduces a chance that we have missed a raw X allocation somewhere in this PR or will miss it somewhere in backporting.


> Overall, it is an improvement that should not be reverted IMO.

IMO simplicity is better API design. The result of `new` should be stored in a `Pointer` - that is a constant for all types in Squid with `Pointer` defined. This is gratuitously forcing developers to work with an unusual (for C++) allocation mechanism.

> 
> > Eduard: the factory style would involve some dynamic type selection
> 
> Let's ignore unimportant terminology differences. The essence of this change request is clear -- replace `X::Make()` with `new X()`. I see no justification for going back to raw pointers, but perhaps I am missing something. In either case, we can save time by _not_ arguing about the definition of "maker-factory style".

Nod. This is not a request to "go back to raw pointers" it is a request to remove a layer of abstraction between the construction and Pointer storage.

File: src/helper/forward.h
Comment (yadij) [reply src/helper/forward.h:30]: I am okay with that naming. Yes.

**Review by yadij — CHANGES_REQUESTED** [review]
Maker and Pointer definitions still need to be resolved.

File: src/helper/forward.h
Comment (eduard-bagdasaryan) [reply src/helper/forward.h]: Done (72c531e).

**Review by yadij — APPROVED** [review]

**Pull request closed without merging**
